src/duk_debugger.c

```diff
@@ -1448,6 +1448,43 @@ DUK_LOCAL void duk__debug_handle_detach(duk_hthread *thr, duk_heap *heap) {
 	DUK__SET_CONN_BROKEN(thr, 0);  /* not an error */
 }
 
+DUK_LOCAL void duk__debug_handle_apprequest(duk_hthread *thr, duk_heap *heap) {
+	DUK_D(DUK_DPRINT("debug command AppRequest"));
+
+	if (heap->dbg_request_cb != NULL) {
+		duk_context *ctx = (duk_context *) thr;
+		duk_idx_t nrets;
+		duk_idx_t nvalues = 0;
+		duk_idx_t old_top;
+		duk_idx_t idx, end;
+
+		/* read tvals from the message and push them onto the valstack,
+		 * then call the request callback to process the request.
+		 */
+		old_top = duk_get_top(ctx);  /* save stack top */
+		while (duk_debug_peek_byte(thr) != DUK_DBG_MARKER_EOM) {
+			duk_debug_read_tval(thr);  /* push to stack */
+			nvalues++;
+		}
+
+		/* Request callback should push values for reply to client onto valstack */
+		nrets = heap->dbg_request_cb(heap->dbg_udata, ctx, nvalues);
+
+		/* Reply with tvals pushed by request callback */
+		duk_debug_write_byte(thr, DUK_DBG_MARKER_REPLY);
+		end = duk_get_top(ctx);
+		for (idx = end - nrets; idx < end; idx++) {
+			duk_debug_write_tval(thr, duk_get_tval(ctx, idx));
```

Comment:
Here again we could use `DUK_GET_TVAL_POSIDX(ctx, idx)`.

However, the `idx` argument could be technically below the valstack bottom here if nrets is insane? Maybe it should be validated (outside an assert)? Then again, we should be able to trust the callback because it can mess things up badly anyway.